lib/Enforcers/RevngEnforcers.cpp

```diff
@@ -0,0 +1,102 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/StringMap.h"
+
+#include "revng/AutoEnforcer/AutoEnforcerLibraryRegistry.h"
+#include "revng/AutoEnforcer/AutoEnforcerTarget.h"
+#include "revng/AutoEnforcer/InspectableLLVMContainer.h"
+#include "revng/Enforcers/CompileModule.h"
+#include "revng/Enforcers/Isolate.h"
+#include "revng/Enforcers/Lift.h"
+#include "revng/Enforcers/LinkForTranslation.h"
+#include "revng/Enforcers/LinkSupport.h"
+#include "revng/Enforcers/RevngEnforcers.h"
+#include "revng/Enforcers/RootKind.h"
+#include "revng/Enforcers/TemporaryFile.h"
+
+using namespace std;
+using namespace AutoEnforcer;
+
+Granularity AutoEnforcer::RootGranularity("Root Granularity");
+Granularity
+  AutoEnforcer::FunctionsGranularity("Function Granularity", RootGranularity);
+
+Kind AutoEnforcer::Binary("Binary", &RootGranularity);
+
+Kind AutoEnforcer::Object("Object", &RootGranularity);
+Kind AutoEnforcer::Translated("Translated", &RootGranularity);
+Kind AutoEnforcer::Dead("Dead", &RootGranularity);
+
+Kind AutoEnforcer::CFepper("CFepper", &FunctionsGranularity);
+Kind AutoEnforcer::ABIEnforced("ABIEnforced", Isolated, &FunctionsGranularity);
+
+llvm::Error RevngModelAutoEnforcerWrapper::save(llvm::StringRef Path) const {
+  return serializeToFile(Model.getReadOnlyModel(), Path);
+}
+
+llvm::Error RevngModelAutoEnforcerWrapper::load(llvm::StringRef Path) {
+  if (not llvm::sys::fs::exists(Path))
+    return llvm::Error::success();
+
+  auto MaybeModel = TupleTree<model::Binary>::loadFromFile(Path);
+  if (not MaybeModel)
+    return llvm::make_error<llvm::StringError>("Could not parse model",
+                                               MaybeModel.getError());
+
+  Model = std::move(*MaybeModel);
+  return llvm::Error::success();
+}
+namespace AutoEnforcer {
+llvm::Error autoEnforcerRevngConfigurationCallback(const PipelineLoader &Loader,
+                                                   LLVMEnforcer &NewPass) {
+  using Wrapper = RevngModelAutoEnforcerWrapper;
+  auto MaybeModelWrapper = Loader.getContext().get<Wrapper>(Wrapper::Name);
+  if (not MaybeModelWrapper)
+    return MaybeModelWrapper.takeError();
+
+  auto &ModelWrapper = (*MaybeModelWrapper)->getModelWrapper();
+  NewPass.emplacePass<LoadModelEnforcerPass>(ModelWrapper);
+  return llvm::Error::success();
+}
+
+const char RevngModelAutoEnforcerWrapper::ID = '0';
+
+class RevngRegistryLibrary : public AutoEnforcerLibraryRegistry {
```

Comment:
Let's have a separate directory for revng as a user of the autoenforcer. Ideally, one day we could distribute `libAutoEnforcer.so` as a revng-agnostic, standalone component.

Also, let's not prefix things with `Revng*`. If you'd like, you're officially allowed to introduce for the first time the `revng` namespace.

Reply:
i can simply drop the revng prefix, stuff that does not depends only on support already is in the enforcer librarys instead of the autoenforcer one